### ISSUE TITLE
qat_hw_ciphers.c: potential buffer overflow based on a size that is derived from external data

### ISSUE THREAD

**Issue by ColinIanKing** [intel/QAT_Engine#270]
In function qat_chained_ciphers_do_cipher there is a potential buffer overflow due to an unchecked for-loop start index value:

```
        memcpy(d_fbuf[1].pData, inb, buflen - discardlen);

        if (enc) {
            /* Add padding to input buffer at end of digest */
            for (i = plen + dlen; i < buflen; i++)
                d_fbuf[1].pData[i] = pad_len;
        } else {
```

the loop index i starts and plen + dlen, however plen is derived from GET_TLS_PAYLOAD_LEN(tls_hdr) - plen_adj which is based on external data:

```
             if (vtls >= TLS1_VERSION)
                plen = GET_TLS_PAYLOAD_LEN(tls_hdr) - plen_adj;

            /* Compute the padding length using total buffer length, payload
             * length, digest length and a byte to encode padding len.
             */
            pad_len = buflen - (plen + dlen) - 1;
```

If the tls_hdr data is corrupted then the index i can end up outside the pData[] buffer leading to a buffer overflow or a denial of service issue.  I suggest variable i requires bounds checking before it is used as an index into pData[].

**Comment by Yogaraj-Alamenda** [comment]
@ColinIanKing Thanks for the inputs. We will fix it in upcoming release

**Comment by Yogaraj-Alamenda** [comment]
Fixed in the commit 418411b5206c2727f6578fbca9dbd11df98a237d
